smplrout.cc

```diff
@@ -211,123 +151,139 @@ void SimplifyRouteFilter::routesimple_head(const route_head* rte)
+  QHash<Waypoint*, trackerror> wpthash;
+  for (auto* wpt : rte->waypoint_list) {
+    wpt->extra_data = nullptr;
+
+    if (metric == metric_t::relative) {
+      // check hdop is available for compute_track_error
+      if (wpt->hdop == 0) {
+        fatal(MYNAME ": relative needs hdop information.\n");
+      }
+    }
 
-}
+    if (pwpt != nullptr) {
+      nb.wpt = pwpt;
+      nb.prev = ppwpt;
+      nb.next = wpt;
 
-void SimplifyRouteFilter::shuffle_xte(struct xte* xte_rec)
-{
-  while (xte_rec > xte_recs && compare_xte(xte_rec, xte_rec-1) < 0) {
-    std::swap(xte_rec[0], xte_rec[-1]);
-    xte_rec[0].intermed->xte_rec = &xte_rec[0];
-    xte_rec[-1].intermed->xte_rec = &xte_rec[-1];
-    xte_rec--;
-  }
-  while (xte_rec - xte_recs < xte_count-2 &&
-         compare_xte(xte_rec, xte_rec+1) > 0) {
-    std::swap(xte_rec[0], xte_rec[1]);
-    xte_rec[0].intermed->xte_rec = &xte_rec[0];
-    xte_rec[1].intermed->xte_rec = &xte_rec[1];
-    xte_rec++;
-  }
-}
+      te.dist = compute_track_error(nb);
+      te.wptpos = pos;
 
-void SimplifyRouteFilter::routesimple_tail(const route_head* rte)
-{
-  int i;
-  if (!cur_rte) {
-    return;
-  }
+      errormap.insert(te, nb);
+      wpthash.insert(pwpt, te);
+    }
 
-  /* compute all distances */
-  for (i = 0; i < xte_count ; i++) {
-    compute_xte(xte_recs+i);
+    ppwpt = pwpt;
+    pwpt = wpt;
+    ++pos;
   }
+  nb.wpt = pwpt;
+  nb.prev = ppwpt;
+  nb.next = nullptr;
 
+  te.dist = compute_track_error(nb);
+  te.wptpos = pos;
 
-  /* sort XTE array, lowest XTE last */
-  auto compare_xte_lambda = [](const xte& a, const xte& b)->bool {
-    return compare_xte(&a, &b) < 0;
-  };
-  if (gpsbabel_testmode()) {
-    std::stable_sort(xte_recs, xte_recs + xte_count, compare_xte_lambda);
-  } else {
-    std::sort(xte_recs, xte_recs + xte_count, compare_xte_lambda);
-  }
+  errormap.insert(te, nb);
+  wpthash.insert(pwpt, te);
 
-
-  for (i = 0; i < xte_count; i++) {
-    xte_recs[i].intermed->xte_rec = xte_recs+i;
-  }
-  // Ensure totalerror starts with the distance between first and second points
-  // and not the zero-init.  From a June 25, 2014  thread titled "Simplify
-  // Filter: GPSBabel removes one trackpoint..."  I never could repro it it
-  // with the sample data, so there is no automated test case, but Steve's
-  // fix is "obviously" right here.
-  if (xte_count >= 1) {
-    totalerror = xte_recs[xte_count-1].distance;
-  }
+  assert(!errormap.isEmpty());
+  double totalerror = errormap.lastKey().dist;
 
   /* while we still have too many records... */
-  while ((xte_count) &&
-         (((limit_basis == limit_basis_t::count) && (count < xte_count)) ||
+  while ((!errormap.isEmpty()) &&
+         (((limit_basis == limit_basis_t::count) && (count < errormap.size())) ||
           ((limit_basis == limit_basis_t::error) && (totalerror < error)))) {
-    i = xte_count - 1;
+
     /* remove the record with the lowest XTE */
+    neighborhood goner = errormap.last();
```

Comment:
I just did 8e6 points -> simplify -> 4e6 points in 48 seconds (for the filter alone), so I don't think we should just stick with the algorithm.  I think this would have taken days with 1.8.0.